File: zig/src/problem0002/add_two_numbers.zig
const std = @import("std");
const ListNode = @import("./linked_list.zig").ListNode;

pub fn add_two_numbers(l1: ?*ListNode, l2: ?*ListNode) ?*ListNode {
    const head = std.heap.page_allocator.create(ListNode) catch unreachable;

    var current = head;
    var l1_ptr = l1;
    var l2_ptr = l2;
    var carry: i32 = 0;

    // * 同时遍历两个链表，逐步计算它们的和，并与当前进位值相加
    // * 遍历结束，如果还有进位值，则在链表末尾添加一个新节点
    while (l1_ptr != null or l2_ptr != null or carry != 0) {
        var sum = carry;

        if (l1_ptr != null) {
            sum += l1_ptr.?.val;
            l1_ptr = l1_ptr.?.next;
        }

        if (l2_ptr != null) {
            sum += l2_ptr.?.val;
            l2_ptr = l2_ptr.?.next;
        }

        current.next = std.heap.page_allocator.create(ListNode) catch unreachable;
        current = current.next.?;
        current.* = .{ .val = @mod(sum, 10) };
        carry = @divFloor(sum, 10);
    }

    return head.next;
}

File: typescript/src/problem0002/addTwoNumbers.ts
import ListNode from "./linkedList";

export default function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
  const head = new ListNode();

  let current = head,
    carry = 0;

  // * 同时遍历两个链表，逐步计算它们的和，并与当前进位值相加
  // * 遍历结束，如果还有进位值，则在链表末尾添加一个新节点
  while (l1 !== null || l2 !== null || carry !== 0) {
    const sum = (l1?.val ?? 0) + (l2?.val ?? 0) + carry;

    l1 = l1?.next ?? null;
    l2 = l2?.next ?? null;

    current.next = new ListNode(sum % 10);
    current = current.next;
    carry = Math.floor(sum / 10);
  }

  return head.next;
}
